### ISSUE TITLE
Goroutine leak in handshake server on version mismatch refusal

### ISSUE THREAD

**Issue by agaffney** [blinklabs-io/gouroboros#535]
With goleak enabled in the handshake server tests, we get the following. Uncomment goleak in the handshake server tests when this is fixed.

```
=== RUN   TestServerHandshakeRefuseVersionMismatch
    server_test.go:152: found unexpected goroutines:
        [Goroutine 193 in state select, with github.com/blinklabs-io/gouroboros/protocol.(*Protocol).stateLoop on top of the stack:
        github.com/blinklabs-io/gouroboros/protocol.(*Protocol).stateLoop(0xc0001ce4e0, 0xc000091aa0)
        	/home/agaffney/dev/blinklabs-io/gouroboros/protocol/protocol.go:438 +0x2ba
        created by github.com/blinklabs-io/gouroboros.(*Connection).setupConnection.(*Protocol).Start.func7 in goroutine 182
        	/home/agaffney/dev/blinklabs-io/gouroboros/protocol/protocol.go:135 +0x3fd
         Goroutine 211 in state select, with github.com/blinklabs-io/gouroboros/protocol.(*Protocol).sendLoop on top of the stack:
        github.com/blinklabs-io/gouroboros/protocol.(*Protocol).sendLoop(0xc0001ce4e0)
        	/home/agaffney/dev/blinklabs-io/gouroboros/protocol/protocol.go:184 +0x185
        created by github.com/blinklabs-io/gouroboros.(*Connection).setupConnection.(*Protocol).Start.func7 in goroutine 182
        	/home/agaffney/dev/blinklabs-io/gouroboros/protocol/protocol.go:137 +0x4d1
        ]
--- FAIL: TestServerHandshakeRefuseVersionMismatch (0.45s)
```

**Comment by igorcrevar** [comment]
The `recvLoop` function for `Protocol` receives error messages through its handler:
```go
if err := p.handleMessage(msg); err != nil {
    p.SendError(err)
    return
}
```

Following this, the `WaitGroup` is decremented with `p.waitGroup.Done()`. Initially, `p.waitGroup.Add(2)` is used to account for both the `recvLoop` and `sendLoop` routines.

However, the issue arises because `sendLoop` is never notified when `recvLoop` quits and continues to run. The same problem occurs with `stateLoop.` 

To address this, a potential easy fix is to modify the `recvLoop` function by using the following defer statement:
```go
defer func() {
    p.waitGroup.Done()
    close(p.doneChan)
}()
```
This modification triggers the shutdown of all three routines. However, it's important to consider potential consequences, as I am uncertain about them. The code may benefit from further refactoring to improve channel handling and clarify responsibilities.

**Comment by agaffney** [comment]
Fixed in #542
